qos-monitor/src/main/java/eu/arrowhead/core/qos/service/ping/monitor/PingEventProcessor.java

```diff
@@ -0,0 +1,147 @@
+package eu.arrowhead.core.qos.service.ping.monitor;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.Resource;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.springframework.stereotype.Service;
+import org.springframework.util.Assert;
+
+import eu.arrowhead.common.exception.ArrowheadException;
+import eu.arrowhead.core.qos.QosMonitorConstants;
+import eu.arrowhead.core.qos.dto.IcmpPingDTOConverter;
+import eu.arrowhead.core.qos.dto.IcmpPingResponse;
+import eu.arrowhead.core.qos.dto.event.monitoringevents.FinishedMonitoringMeasurementEventDTO;
+import eu.arrowhead.core.qos.dto.event.monitoringevents.InterruptedMonitoringMeasurementEventDTO;
+import eu.arrowhead.core.qos.dto.event.monitoringevents.MeasurementMonitoringEvent;
+import eu.arrowhead.core.qos.dto.event.monitoringevents.ReceivedMonitoringRequestEventDTO;
+import eu.arrowhead.core.qos.dto.event.monitoringevents.StartedMonitoringMeasurementEventDTO;
+
+@Service
+public class PingEventProcessor {
+	//=================================================================================================
+	// members
+
+	private static final long SLEEP_PERIOD = TimeUnit.SECONDS.toMillis(1);
+	private static final String SUCCESS_FINISH_PHRASE = " successfully finished ";
+	private static final String INTERRRUPTED_FINISH_PHRASE = " get interrupted";
+	private static final String TIMEOUT_FINISH_PHRASE = " timed out";
+
+	@Resource(name = QosMonitorConstants.EVENT_BUFFER)
+	private ConcurrentHashMap<UUID, PingEventBufferElement> eventBuffer;
+
+	private final Logger logger = LogManager.getLogger(PingEventProcessor.class);
+
+	//=================================================================================================
+	// methods
+
+	//-------------------------------------------------------------------------------------------------	
+	public List<IcmpPingResponse> processEvents(final UUID id, final long measurementExpiryTime) {
+		logger.debug("processEvents started...");
+
+		Assert.notNull(id, "Event id could not be null.");
+		Assert.isTrue(measurementExpiryTime > 0, "ExpieryTime should be greater than zero.");
+		Assert.isTrue(measurementExpiryTime > System.currentTimeMillis(), "ExpieryTime should be in the future.");
+
+		ReceivedMonitoringRequestEventDTO temporalReceivedRequestEvent = null;
+		StartedMonitoringMeasurementEventDTO temporalStartedMonitoringEvent = null;
+
+		while ( System.currentTimeMillis() < measurementExpiryTime) {
+
+			final PingEventBufferElement element = eventBuffer.get(id);
+			if (element == null) {
+
+				rest();
+				continue;
+
+			}else if (Arrays.asList(element.getEventArray()).isEmpty()){
```

Comment:
it causes NPE if element.getEventArray() is null. So maybe use this:

`element.getEventArray() == null || element.getEventArray().length == 0`

Reply:
the eventArray gets created at constract time in the element(bufferElement), what I try to check here is that the eventArray has any non null element.

Reply:
added new method to check if array has only null elements